src/components/ModalAuth.tsx

```diff
@@ -0,0 +1,214 @@
+
+    updateIsSubmitting(true)
+    try {
+      await window.firebase.auth().createUserWithEmailAndPassword(email, password!)
+      closeModal()
+      updateIsSubmitting(false)
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      return updateError(firebaseErrorsIndex[error?.code as errorCode] || firebaseErrorsIndex.default)
+    }
+  }, [])
+
+  const submitAction = useRef<SubmitAction>(signUp)
+
+  /** Handle email change. */
+  const onChangeEmail = (e: ChangeEvent<HTMLInputElement>) => {
+    updateEmail(e.target.value)
+  }
+
+  /** Handle password change. */
+  const onChangePassword = (e: ChangeEvent<HTMLInputElement>) => updatePassword(e.target.value)
+
+  /** Reset password using reset email. */
+  const resetPassword: SubmitAction = useCallback(async (closeModal, email) => {
+    updateIsSubmitting(true)
+
+    try {
+      await window.firebase.auth().sendPasswordResetEmail(email, { url: FIREBASE_REDIRECT_URL! })
+      updateIsSubmitting(false)
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      return updateError(error.message || firebaseErrorsIndex.default)
+    }
+    dispatch(alert('Please check your email'))
+    closeModal()
+  }, [])
+
+  /** Login with email and password. */
+  const loginWithEmailAndPassword: SubmitAction = useCallback(async (closeModal, email, password) => {
+    updateIsSubmitting(true)
+    try {
+      await window.firebase.auth().signInWithEmailAndPassword(email, password!)
+      updateIsSubmitting(false)
+      closeModal()
+    }
+    catch (error) {
+      updateIsSubmitting(false)
+      if (error?.code === 'auth/wrong-password') {
+        updateShowForgotPasswordOption(true)
+      }
+      return updateError(firebaseErrorsIndex[error?.code as errorCode] || firebaseErrorsIndex.default)
+    }
+  }, [])
+
+  /**
+   * Reset Email and Password fields.
+   */
+  const resetFormFields = () => {
+    updateEmail('')
+    updatePassword('')
+    updateError(null)
+  }
+
+  /**
+   * Login using google account.
+   */
+  const signInWithGoogle = () => {
+    updateIsSubmitting(true)
+    dispatch(login())
+  }
+
+  /**
+   * Enable Login with email and password.
+   */
+  const handleLogin = () => {
+    updateModalKey(modalKeys.login)
+    updateTitle(modalTitles.login)
+    updateShowLogin(true)
+    resetFormFields()
+    submitAction.current = loginWithEmailAndPassword
+  }
+
+  /**
+   * Enable Sign up.
+   */
+  const handleSignUp = () => {
+    updateModalKey(modalKeys.signup)
+    updateTitle(modalTitles.signup)
+    updateShowLogin(false)
+    resetFormFields()
+    submitAction.current = signUp
+  }
+
+  /**
+   * Enable Password recovery assistance.
+   */
+  const handleForgotPassword = () => {
+    updateModalKey(modalKeys.resetPassword)
+    updateTitle(modalTitles.resetPassword)
+    updateForgotPassword(true)
+    resetFormFields()
+    submitAction.current = resetPassword
+  }
```

Comment:
This section can be improved a bit. The problem is that the variables are treated as independent when they aren't: they should all flow from the active mode. With separate variables it is possible to update some and not others and thus is more error-prone. 

Instead, store a single local state `active` or `activeMode` which is set to `login`, `signup`, or `resetPassword`. Then for example the title is just `modalTitles[active]` and does not require additional state.

I like the `modalKeys`, `modalTitles`, etc structure. It will enable clean computed values with minimal state.

Reply:
Agreed